File: fabric-mod/src/main/java/net/spritehack/module/Module.java
package net.spritehack.module;

import net.minecraft.client.MinecraftClient;
import net.minecraft.sound.SoundEvents;

public abstract class Module {

    public enum Category {
        COMBAT("Combat"),
        MOVEMENT("Movement"),
        RENDER("Render"),
        PLAYER("Player"),
        MACROS("Macros"),
        CLIENT("Client");

        public final String display;
        Category(String display) { this.display = display; }
    }

    protected String name;
    protected String description;
    protected Category category;
    protected boolean enabled;
    protected int keyBind;

    public Module(String name, String description, Category category) {
        this.name = name;
        this.description = description;
        this.category = category;
        this.enabled = false;
        this.keyBind = -1;
    }

    public void onEnable() {}
    public void onDisable() {}
    public void onTick(MinecraftClient client) {}

    protected long lastToggleTime = 0;

    public void toggle() {
        long now = System.currentTimeMillis();
        if (now - lastToggleTime < 200) return; // 0.2s cooldown
        lastToggleTime = now;

        enabled = !enabled;
        if (enabled) onEnable();
        else onDisable();

        // Audio feedback on toggle
        MinecraftClient mc = MinecraftClient.getInstance();
        if (mc != null && mc.player != null) {
            mc.player.playSound(SoundEvents.UI_BUTTON_CLICK.value(), 0.5f, enabled ? 1.3f : 0.8f);
        }
    }

    public void setEnabled(boolean val) {
        if (val != enabled) toggle();
    }

    public String getName() { return name; }
    public String getDescription() { return description; }
    public Category getCategory() { return category; }
    public boolean isEnabled() { return enabled; }
    public int getKeyBind() { return keyBind; }
    public void setKeyBind(int key) { this.keyBind = key; }
}

File: src/client/Module.js
// Base Module Class for SpriteHack Client
export const CATEGORY = {
  COMBAT: 'Combat',
  MOVEMENT: 'Movement',
  RENDER: 'Render',
  PLAYER: 'Player',
  CLIENT: 'Client'
};

export class Module {
  constructor(name, category, description = '', defaultKey = null) {
    this.name = name;
    this.category = category;
    this.description = description;
    this.keybind = defaultKey;
    this.enabled = false;
    this.settings = [];
  }

  addSetting(setting) {
    this.settings.push(setting);
    return setting;
  }

  toggle() {
    this.setEnabled(!this.enabled);
    return this.enabled;
  }

  setEnabled(state) {
    if (this.enabled === state) return;
    this.enabled = state;
    if (this.enabled) {
      this.onEnable();
    } else {
      this.onDisable();
    }
  }

  onEnable() {}
  onDisable() {}
  onTick(gameContext) {}
  onRender3D(gameContext) {}
  onRender2D(gameContext) {}
}
